### ISSUE TITLE
一些关于代码和论文描述的问题

### ISSUE THREAD

**Issue by JunpoYu** [ozg2021/DRUM#1]
作者您好，我读了您的题为 ["DRUM: Diffusion-based runoff model for probabilistic flood forecasting"](https://arxiv.org/pdf/2412.11942) 的文章，并阅读了相关代码，我有几个问题希望能够向您请教一下。

### 1.  Static attributes 是否和 contidion 的描述存在矛盾

从您的论文描述中（Page 26）

*The unconditional model captures the general empirical distribution of runoff sequences, whereas the conditional model refines predictions by incorporating additional contextual information, such as past meteorological conditions, static catchment attributes, and current weather states.*

unconditional model 应该只包含原始数据，例如 x0，而其他数据则是作为 condition。但是在您的代码实现中 `epsilon_unconditional.py` 的 Line 82 使用了 xs 作为输入，根据注释这是静态数据  static catchment attribute ，应该是 condition 的一部分。

并且在另一处描述中 (Page 13)

*To build a regional model, we first develop an unconditional diffusion model that incorporates basin’s static attributes* 

则认为 basin’s static attributes 并非是 condition，似乎前后矛盾。

### 2. epsilon_conditional 的输入数据和论文描述

> 非常荣幸我提前阅读过您的 GRL_Manuscript（感谢潘老师提供给我们的学习机会），其中 Figure.1 详细描述了 DRUM 的模型结构，但是其代码实现似乎和图像存在一定出入。

#### 2.1 输入数据个数

首先根据 Figure.1 的描述，conditional model 应该包含五个输入，分别是 `xt, Historical forcings (H), Static attributes (S), Nosie level step (t), Meteorological forecats (F)`，但是在代码实现中仅仅接受了 4 个输入。

```python 
class EpsModel(nn.Module):
    """
    EpsModel: A conditional diffusion model for noise prediction.

    This model combines temporal, static, and forcing information to predict noise
    in the diffusion process. It uses LSTM-based encoders and MLPs for feature
    extraction and integration.
    """


    def forward(self, xt, t, xd, xs): 
        """
        Forward pass through the diffusion noise prediction model.

        Args:
            xt (torch.Tensor): Current noisy state at diffusion time step t.
                              Shape: (batch_size, 1). Represents the noisy target variable.
            t (torch.Tensor): Diffusion time step values for time embedding.
                             Shape: (batch_size,). Values typically in [0, n_steps-1].
            xd (torch.Tensor): Input sequence data (meteorological forcing).
                              Shape: (batch_size, seq_len, input_size).
                              Contains historical meteorological observations.
            xs (torch.Tensor): Static basin attributes (topography, soil, etc.).
                              Shape: (batch_size, static_attr_len).
                              Time-invariant characteristics of the watershed.

        Returns:
            torch.Tensor: Predicted noise at the current diffusion step.
                         Shape: (batch_size, future_step).
                         Used to denoise xt in the reverse diffusion process.
        """
        batch_size, seq_len, _ = xd.size()

        t_emb = self.time_embedding(t)
        xs_emb = self.static_mlp(xs) 
        xs_emb_expanded = xs_emb.unsqueeze(1).expand(batch_size, seq_len, -1)
        xd_combined = torch.cat((xd, xs_emb_expanded), dim=2)
        xd_emb = self.forcing_embedding(xd_combined) 
        combined_emb = torch.cat((t_emb, xd_emb, xs_emb), dim=1)
        combined_input = torch.cat((xt, t_emb, xd_emb, xs_emb), dim=1) 
        decoder_input = combined_input.unsqueeze(1)
        # print(decoder_input.shape)

        H = torch.zeros(self.num_layers, batch_size, self.hidden_size).to(xt.device)
        C = torch.zeros(self.num_layers, batch_size, self.hidden_size).to(xt.device)

        out, _ = self.decoder_lstm(decoder_input, (H, C)) 
        lstm_output = out.squeeze(1)
        mlp_output = self.mlp(combined_input, combined_emb)
        combined_out = torch.cat((lstm_output, mlp_output), dim=1)
        out = self.fc(combined_out)

        return out
```

`forward(self, xt, t, xd, xs)` 函数中 `xd` 被描述为 historical meteorological observations，即 `H`，且后续使用了 LSTM 进行 embed 操作。但是我没有找到 `meteorological forecasts` 也就是 `F`。如果 `F` 是作为预测（采样）阶段的输入，则 Figure.1 可能存在误导性。或者我没有正确地理解数据构建过程，希望作者能帮我指出 meteorological forecasts 是如何进入模型的。

> 难道 xd 中同时包含了 historical forcings 和 meteorological forecasts？

> PS: 我同样阅读了 `get_predict_ddim.py` 和 `DDIM.ipynb`，由于其中似乎也没有涉及到 meteorological forecasts 的输入。

### 3. 静态数据的嵌入方法

根据您 GRL_manuscript 中 Figure.1 的描述，Static attributes `S` 应该和 Historical forcings `F` 一起经过 LSTM encoder 处理，但是无论是 conditional model 还是 unconditional model 的实现，`S` 都是使用了一个额外的 mlp 进行处理，这一部分是否需要在 Figure.1 中指明？

### 4. 重复拼接

在 `epsilon_conditional.py` 的实现中，有一段比较复杂的数据拼接过程。

```python
      t_emb = self.time_embedding(t)
      xs_emb = self.static_mlp(xs)
      xs_emb_expanded = xs_emb.unsqueeze(1).expand(batch_size, seq_len, -1)
      xd_combined = torch.cat((xd, xs_emb_expanded), dim=2)
      xd_emb = self.forcing_embedding(xd_combined)
      combined_emb = torch.cat((t_emb, xd_emb, xs_emb), dim=1)
      combined_input = torch.cat((xt, t_emb, xd_emb, xs_emb), dim=1)
      decoder_input = combined_input.unsqueeze(1)
```

其中 `xd_emb` 已经包含了 `xs_emb` 的信息，而 `combined_emb` 又一次将 `xs_emb, xd_emb` 拼接到一起，同样的情况也发生在 `combined_input ` 上。这里重复包含 `xs` 的信息是必要的吗？如果是则其拼接过程似乎与 Figure.1 的描述不符。并且 `combined_emb` 和 `combined_input` 一起作为了后续 mlp 的输入，也就是说这个过程中 `t_emb, xd_emb, xs_emb` 都有多条路径重复进入 mlp。

对于以上问题我的理解是，在不同的嵌入过程中使用其他信息作为辅助是必要的，例如处理 `xd` 时给出 `xs` 将会提高模型对于数据的理解能力。而重复性地输入是为了增强条件信息，例如 mlp 中也使用了多次条件注入操作。如果我的理解有误还请您指出。

### 5. 为什么使用 LSTM

在构建 `decoder_input` 时（同上一问代码），实际上这个变量并不是 seq 类型的数据，其 `seq_len` 维度大小为 1，对于这样的数据为什么要使用 LSTM 处理？一个长度为 1 的序列经过 LSTM 和经过一个 mlp 似乎没有什么区别。还是说为了保证符合 LSTM encoder-decoder 架构的描述？

### 6. 关于 LSTM-p 和 LSTM-d 的不确定性

论文中提到

*The probabilistic models (DRUM and LSTM-p) generate 50 ensemble forecasts per time step to characterize prediction uncertainty, whereas the deterministic model (LSTM-d) produces single-valued outputs.*

很遗憾代码中似乎没有给出 LSTM-p 和 LSTM-d 的具体实现。

LSTM-p 给出的是概率分布，为什么还要通过 ensemble 的形式获得其不确定性。LSTM-d 仅仅给出一个值，其不确定性如果不通过 ensemble 又是如何给出的？

### 7. 关于扩散模型的层次建模解释

论文中提到

*DRUM’s iterative refinement process naturally decomposes the complex task of flood prediction into a hierarchy of more manageable sub-problems. The forward process progressively adds noise to the data, while the reverse process learns to reconstruct hydrological patterns at multiple temporal scales. This hierarchical decomposition is evident in the basin-specific CDFs shown in Fig. 7c, where DRUM successfully captures distinct distributional characteristics across diverse hydroclimatic gradients, from arid to humid regions. Early denoising steps recover broad seasonal and inter-annual patterns, while intermediate steps reconstruct medium-term weather responses. Final steps focus
on capturing high-frequency dynamics and extreme events. This hierarchical approach enables more robust learning compared to LSTM models, which must simultaneously capture all temporal scales in a single prediction step. *

Fig.7c 给出了几个干湿条件不同的点的 CDF 图像，如何从这几个图像证明 *Early denoising steps recover broad seasonal and inter-annual patterns, while intermediate steps reconstruct medium-term weather responses. Final steps focus on capturing high-frequency dynamics and extreme events.*

我略微了解有一些关于 diffusion model 的层次建模的探索，因此我比较认可这一段的描述（或者说猜测？），但是 Fig.7c 似乎不足以作为证据？

**Comment by ozg2021** [comment]
**Issue 5:**
The decoder LSTM was designed for the multi-step forecasting task, where it processes sequences of meteorological forecasts. In the nowcasting case (0-day lead time), however, the decoder input reduces to a single timestep, which makes the LSTM appear redundant.

To address this, we added a parallel MLP branch alongside the decoder LSTM. The two branches complement each other: the LSTM preserves consistency with the encoder–decoder architecture required for forecasting, while the MLP efficiently handles single-timestep inputs in nowcasting. Their outputs are combined to leverage the strengths of both.

In the forecasting setting with recursive multi-step predictions, the decoder LSTM becomes essential because it processes forecast sequences step by step and maintains temporal dependencies across future timesteps.

**Comment by ozg2021** [comment]
> Thank you very much for your reply. I'm pleased to hear about your paper's publication in GRL.
> 
> Regarding **issue 5,** I still have a question. You mentioned that the lstm-decoder is designed for multi-step predictions. However, looking at your EpsModel code implementation, it's evident that the specific number of future days to predict is determined by the future_step parameter, which merely dictates the number of neurons in the final Fully Connected (FC) output layer.
> 
> This implies that whether it's nowcasting (0-day lead time) or multi-step forecasting, the decision isn't made by the lstm-decoder, but by the output FC layer. In fact, the line decoder_input = combined_input.unsqueeze(1) and the operations preceding it already limit the lstm-decoder to inherently accept data with a sequence length of seq_len=1. Regardless of whether it's nowcasting or multi-step forecasting, the data entering the lstm-decoder invariably has the shape (batch_size, seq_len=1, input_size). Furthermore, there is no code implementation for the lstm-decoder performing recursive multi-step iterative predictions. Based on the nature of LSTM implementations, its decoded output lstm_output is also a single-timestep piece of data (containing only the output from the last timestep).
> 
> In summary, the lstm-decoder does not appear to implement its intended sequence prediction functionality. It merely processes data of shape (batch_size, seq_len=1, input_size) and outputs a result of shape (batch_size, seq_len=1, output_size). The actual multi-step prediction task is handled by the FC layer, controlled by the future_step parameter, which directly outputs predictions for future_step steps ahead. Therefore, your statement: "_In the forecasting setting with recursive multi-step predictions, the decoder LSTM becomes essential because it processes forecast sequences step by step and maintains temporal dependencies across future timesteps_" does not seem to be correctly implemented in the code.
> 
> This is my understanding of the code. If there are any inaccuracies in my interpretation, please kindly point them out. Thank you for taking the time to reply amidst your busy schedule!

Thank you for the detailed analysis. You're correct that the code shown doesn't explicitly demonstrate the recursive generation. The released code is indeed a simplified version optimized for the nowcasting case.

For multi-step forecasting, the decoder LSTM would operate recursively. Here's the **pseudocode** showing how this would work:

```python
# Recursive multi-step forecasting (pseudocode)
h_n, c_n = encoder_final_state()              # Encoder's final hidden/cell states
decoder_state = prepare_decoder_input()       # Contains x_t, diffusion_t_embedding, static_embedding
predictions = []

for k in range(future_step):
    forcing_k = get_forecast_forcing(k)       # Meteorological forecast at horizon k
    decoder_input = concat(decoder_state, forcing_k)

    output, (h_n, c_n) = decoder_lstm(decoder_input, (h_n, c_n))

    hidden_repr = extract_hidden_features(h_n)
    pred_k = output_projection(concat(hidden_repr, encoder_context))
    predictions.append(pred_k)

    decoder_state = update_decoder_state(pred_k, k+1)

lstm_output = stack(predictions)  # Shape: (batch_size, future_step)
```

The key difference is that the hidden state `(h_n, c_n)` carries temporal information across prediction steps, and each step's output feeds into the next (autoregressive generation). Without this recursive loop, the decoder LSTM in the simplified code indeed appears redundant. We released the simplified version for clarity but should have clarified that the full forecasting implementation involves this recursive process.

Thank you again for your careful review. Please feel free to reach out if you have any further questions.
